### PR TITLE
Solution to Mars Rover with test

### PR DESCRIPTION
**Pull request by PriyangaPKini**

### CODE REVIEW

File: src/mars_rover/core.clj [hunk 1]
```diff
@@ -1,6 +1,85 @@
...
+(ns mars-rover.core
+  (:require [clojure.string :as string]))
 
-(defn foo
-  "I don't do a whole lot."
-  [x]
-  (println x "Hello, World!"))
+(def file-path "/Users/priyangapkini/Clojure/mars-rover/mars-rover/src/mars_rover/input.txt")
+
+(defn read-file
```
Comment (jysandy) [inline src/mars_rover/core.clj:6]: This function is doing too many things. It's reading the file, and splitting it into lines. As discussed over call, we can instead inline this and use the `->` macro.

File: src/mars_rover/core.clj [hunk 2]
```diff
@@ -1,6 +1,85 @@
...
+  (let [upper-x (first upper-corner) upper-y (second upper-corner)]
+    (if (not (and (<= 0 curr-x upper-x) (<= 0 curr-y upper-y)))
+      (throw (Exception. "This instruction cannot be excuted. Rover is at the edge of the plateau!!"))
+      state)))
+
+(defn move
+  "Moves rover by one grid w.r.t the orientation"
+  [[x y orientation]]
+  (condp = orientation
+    "N" (is-with-in-boundry? [x (+ y 1) orientation])
```
Comment (jysandy) [inline src/mars_rover/core.clj:39]: Checking whether the new state is within the boundary should be done in this function.

File: src/mars_rover/core.clj [hunk 3]
```diff
@@ -1,6 +1,85 @@
...
+  (condp = orientation
+    "N" [x y "E"]
+    "E" [x y "S"]
+    "W" [x y "N"]
+    "S" [x y "W"]
+    (str "Invalid orientation, \"" orientation \")))
+
+(defn execute
+  "Returns the position of the rover after executing an instruction"
+  [[x y orientation :as state] instruction]
```
Comment (jysandy) [inline src/mars_rover/core.clj:67]: The rover state should be modelled as a map, not a vector.

File: test/mars_rover/core_test.clj [hunk 4]
```diff
@@ -1,7 +1,69 @@
 (ns mars-rover.core-test
   (:require [clojure.test :refer [deftest testing is]]
-            [mars-rover.core]))
+            [mars-rover.core :as mars-rover]))
+
+(def root-path "/Users/priyangapkini/Clojure/mars-rover/mars-rover/src/mars_rover/")
```
Comment (jysandy) [inline test/mars_rover/core_test.clj:5]: We should use a relative path here. These tests will only run on Priyanga's computer.

File: src/mars_rover/input.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+5 5
```
Comment (jysandy) [inline src/mars_rover/input.txt:1]: This file should be moved to `resources`, it shouldn't be in the `src` folder.

File: src/mars_rover/input1.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (jysandy) [inline src/mars_rover/input1.txt:1]: This file should be moved to `resources`, it shouldn't be in the src folder.

**Review by jysandy — CHANGES_REQUESTED** [review]
I had a long discussion with @PriyangaPKini explaining in detail the changes to be made. I've left some brief comments as a general indicator of what should be changed.

File: src/mars_rover/core.clj (hunk 3 above)
Comment (ssrihari) [reply src/mars_rover/core.clj:65]: "state" is not a good name, it doesn't represent what it holds, and is a generic word like `data`, `details`, `info`, etc.

File: src/mars_rover/core.clj [hunk 7]
```diff
@@ -1,6 +1,82 @@
...
+  (condp = orientation
+    "N" [x y "E"]
+    "E" [x y "S"]
+    "W" [x y "N"]
+    "S" [x y "W"]))
+
+(defn final-position
+  "Returns the position of the rover after executing an instruction"
+  [[x y orientation :as state] instruction]
+  (let [x (to-int x) y (to-int y)]
```
Comment (ssrihari) [inline src/mars_rover/core.clj:66]: Separate the concerns of parsing input from actually moving the rover.

File: src/mars_rover/core.clj (hunk 3 above)
Comment (ssrihari) [reply src/mars_rover/core.clj:65]: this function is not returning the "final" position.. it returns the position after processing **one** instruction.

File: src/mars_rover/core.clj [hunk 8]
```diff
@@ -1,6 +1,82 @@
...
+  "Returns the position of the rover after executing a sequence of instructions"
+  []
+  (loop [index 1]
+    (when (< index (count inputs))
+      (try
+        (prn (reduce final-position
+                     (string/split (nth inputs index) #" ")
+                     (seq (nth inputs (inc index)))))
+        (catch Exception e (str "caught exception: " (.getMessage e))))
+      (recur (+ 2 index)))))
```
Comment (ssrihari) [inline src/mars_rover/core.clj:82]: Do you really need the loop?

File: src/mars_rover/core.clj [hunk 9]
```diff
@@ -1,6 +1,82 @@
...
+(defn plateau-explore
+  "Returns the position of the rover after executing a sequence of instructions"
+  []
+  (loop [index 1]
+    (when (< index (count inputs))
+      (try
+        (prn (reduce final-position
+                     (string/split (nth inputs index) #" ")
+                     (seq (nth inputs (inc index)))))
+        (catch Exception e (str "caught exception: " (.getMessage e))))
```
Comment (ssrihari) [inline src/mars_rover/core.clj:81]: Alignment: `(catch Exception e` in one line, the body of the `catch` in another line.

**Review by ssrihari — CHANGES_REQUESTED** [review]

File: test/mars_rover/core_test.clj [hunk 10]
```diff
@@ -88,3 +89,9 @@
       (is (= false (rover/within-boundary?
                     {:x 2 :y 6 :orientation :W}
                     {:max-x 5 :max-y 5}))))))
+
+(deftest test-explore-plateau
+  (testing "when rover is within the plateau while executing the instructions"
+    (is (= [{:x 1, :y 3, :orientation :N} {:x 5, :y 1, :orientation :E}]
+           (mars-rover/explore-plateau (str root-path "input.txt")))))
+  (testing "when rover is taken to the edge of the plateau while executing the instructions"))
```
Comment (PriyangaPKini) [inline test/mars_rover/core_test.clj:97]: Will be updated soon after fixing the exception handling

File: src/mars_rover/core.clj [hunk 11]
```diff
@@ -16,6 +16,9 @@
                           parser/parse-coordinates
                           (zipmap [:max-x :max-y]))
         raw-input     (rest input-lines)]
-    (map #(mars-rover/get-final-position plateau-size %)
-         (parser/parse-state-and-instructions raw-input))))
+    (try
+      (map #(mars-rover/get-final-position plateau-size %)
+           (parser/parse-state-and-instructions raw-input))
+      (catch Exception e
```
Comment (PriyangaPKini) [inline src/mars_rover/core.clj:22]: Need to fix exception handling here. Not able to catch the error I am throwing from other function.

**Pull request merged**